File: backend_mapato/app/Models/Communication.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Carbon\Carbon;

class Communication extends Model
{
    use HasFactory;

    /**
     * The attributes that are mass assignable.
     */
    protected $fillable = [
        'driver_id',
        'driver_name',
        'message_date',
        'message_content',
        'response',
        'mode',
    ];

    /**
     * The attributes that should be cast.
     */
    protected $casts = [
        'message_date' => 'datetime',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Communication modes enumeration
     */
    const MODES = [
        'sms' => 'SMS',
        'call' => 'Simu',
        'whatsapp' => 'WhatsApp',
        'system_note' => 'Kumbuka ya Mfumo'
    ];

    /**
     * Get the driver that owns the communication.
     */
    public function driver()
    {
        return $this->belongsTo(Driver::class);
    }

    /**
     * Check if communication has a response
     */
    public function getHasResponseAttribute(): bool
    {
        return !empty($this->response);
    }

    /**
     * Get truncated message content for display
     */
    public function getTruncatedContentAttribute(): string
    {
        return strlen($this->message_content) > 50 
            ? substr($this->message_content, 0, 50) . '...' 
            : $this->message_content;
    }

    /**
     * Get truncated response for display
     */
    public function getTruncatedResponseAttribute(): string
    {
        if (!$this->has_response) {
            return 'Hakuna jibu';
        }
        
        return strlen($this->response) > 50 
            ? substr($this->response, 0, 50) . '...' 
            : $this->response;
    }

    /**
     * Get formatted message date
     */
    public function getFormattedMessageDateAttribute(): string
    {
        return $this->message_date->format('d/m/Y');
    }

    /**
     * Get formatted date and time
     */
    public function getFormattedDateTimeAttribute(): string
    {
        return $this->message_date->format('d/m/Y H:i');
    }

    /**
     * Get display name for communication mode
     */
    public function getModeDisplayNameAttribute(): string
    {
        return self::MODES[$this->mode] ?? ucfirst($this->mode);
    }

    /**
     * Get icon for communication mode
     */
    public function getModeIconAttribute(): string
    {
        switch ($this->mode) {
            case 'sms':
                return '📱';
            case 'call':
                return '📞';
            case 'whatsapp':
                return '📲';
            case 'system_note':
                return '📝';
            default:
                return '💬';
        }
    }

    /**
     * Scope to get unanswered communications
     */
    public function scopeUnanswered($query)
    {
        return $query->whereNull('response')->orWhere('response', '');
    }

    /**
     * Scope to get answered communications
     */
    public function scopeAnswered($query)
    {
        return $query->whereNotNull('response')->where('response', '!=', '');
    }

    /**
     * Scope to get recent communications (within specified days)
     */
    public function scopeRecent($query, $days = 7)
    {
        return $query->where('message_date', '>=', Carbon::now()->subDays($days));
    }

    /**
     * Scope to filter by communication mode
     */
    public function scopeByMode($query, $mode)
    {
        return $query->where('mode', $mode);
    }

    /**
     * Scope to search by content or driver name
     */
    public function scopeSearch($query, $search)
    {
        return $query->where(function ($q) use ($search) {
            $q->where('driver_name', 'like', "%{$search}%")
              ->orWhere('message_content', 'like', "%{$search}%")
              ->orWhere('response', 'like', "%{$search}%");
        });
    }

    /**
     * Get communication summary statistics
     */
    public static function getSummaryStats()
    {
        $total = self::count();
        $unanswered = self::unanswered()->count();
        $recent = self::recent()->count();
        
        $byMode = [];
        foreach (array_keys(self::MODES) as $mode) {
            $byMode[$mode] = self::byMode($mode)->count();
        }
        
        $lastCommunication = self::latest('message_date')->first();
        
        return [
            'total_communications' => $total,
            'unanswered_communications' => $unanswered,
            'recent_communications' => $recent,
            'communications_by_mode' => $byMode,
            'last_communication_date' => $lastCommunication ? $lastCommunication->message_date : null,
            'response_rate' => $total > 0 ? round((($total - $unanswered) / $total) * 100) : 0
        ];
    }
}

File: boda_mapato/lib/models/communication.dart
class Communication {
  Communication({
    required this.driverId,
    required this.driverName,
    required this.messageDate,
    required this.messageContent,
    required this.mode,
    required this.createdAt,
    required this.updatedAt,
    this.id,
    this.response,
  });

  // Factory constructor from JSON
  factory Communication.fromJson(Map<String, dynamic> json) {
    return Communication(
      id: json["id"],
      driverId: json["driver_id"],
      driverName: json["driver_name"] ?? "",
      messageDate: DateTime.parse(json["message_date"]),
      messageContent: json["message_content"] ?? "",
      response: json["response"],
      mode: CommunicationMode.fromString(json["mode"] ?? "system_note"),
      createdAt: DateTime.parse(json["created_at"]),
      updatedAt: DateTime.parse(json["updated_at"]),
    );
  }
  final int? id;
  final String driverId;
  final String driverName; // Denormalized for easier display
  final DateTime messageDate;
  final String messageContent;
  final String? response;
  final CommunicationMode mode;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Convert to JSON
  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "driver_id": driverId,
      "driver_name": driverName,
      "message_date": messageDate.toIso8601String(),
      "message_content": messageContent,
      "response": response,
      "mode": mode.value,
      "created_at": createdAt.toIso8601String(),
      "updated_at": updatedAt.toIso8601String(),
    };
  }

  // Create copy with updated fields
  Communication copyWith({
    int? id,
    String? driverId,
    String? driverName,
    DateTime? messageDate,
    String? messageContent,
    String? response,
    CommunicationMode? mode,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Communication(
      id: id ?? this.id,
      driverId: driverId ?? this.driverId,
      driverName: driverName ?? this.driverName,
      messageDate: messageDate ?? this.messageDate,
      messageContent: messageContent ?? this.messageContent,
      response: response ?? this.response,
      mode: mode ?? this.mode,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Formatted date string for display
  String get formattedMessageDate {
    return "${messageDate.day.toString().padLeft(2, '0')}/"
        "${messageDate.month.toString().padLeft(2, '0')}/"
        "${messageDate.year}";
  }

  // Formatted time string for display
  String get formattedMessageTime {
    return "${messageDate.hour.toString().padLeft(2, '0')}:"
        "${messageDate.minute.toString().padLeft(2, '0')}";
  }

  // Full formatted datetime for display
  String get formattedDateTime {
    return "$formattedMessageDate $formattedMessageTime";
  }

  // Check if communication has response
  bool get hasResponse {
    return response != null && response!.isNotEmpty;
  }

  // Get truncated content for table display
  String get truncatedContent {
    if (messageContent.length <= 50) return messageContent;
    return "${messageContent.substring(0, 47)}...";
  }

  // Get truncated response for table display
  String get truncatedResponse {
    if (!hasResponse) return "Hakuna jibu";
    if (response!.length <= 30) return response!;
    return "${response!.substring(0, 27)}...";
  }

  @override
  String toString() {
    return 'Communication{id: $id, driverId: $driverId, driverName: $driverName, messageDate: $messageDate, mode: ${mode.displayName}}';
  }
}

// Enum for communication modes
enum CommunicationMode {
  sms("sms", "SMS", "📱"),
  call("call", "Simu", "📞"),
  whatsapp("whatsapp", "WhatsApp", "💬"),
  systemNote("system_note", "Kumbuka za Mfumo", "📝");

  const CommunicationMode(this.value, this.displayName, this.icon);

  final String value;
  final String displayName;
  final String icon;

  // Create from string value
  static CommunicationMode fromString(String value) {
    switch (value.toLowerCase()) {
      case "sms":
        return CommunicationMode.sms;
      case "call":
        return CommunicationMode.call;
      case "whatsapp":
        return CommunicationMode.whatsapp;
      case "system_note":
      default:
        return CommunicationMode.systemNote;
    }
  }

  // Get all modes as a list
  static List<CommunicationMode> get allModes => CommunicationMode.values;
}

// Communication summary for dashboard/overview
class CommunicationSummary {
  CommunicationSummary({
    required this.totalCommunications,
    required this.unansweredCommunications,
    required this.recentCommunications,
    required this.communicationsByMode,
    this.lastCommunicationDate,
  });

  factory CommunicationSummary.fromJson(Map<String, dynamic> json) {
    // Parse communications by mode
    final Map<CommunicationMode, int> modeMap = {};
    if (json["communications_by_mode"] != null) {
      final Map<String, dynamic> modeData = json["communications_by_mode"];
      for (final String key in modeData.keys) {
        final CommunicationMode mode = CommunicationMode.fromString(key);
        modeMap[mode] = modeData[key] ?? 0;
      }
    }

    return CommunicationSummary(
      totalCommunications: json["total_communications"] ?? 0,
      unansweredCommunications: json["unanswered_communications"] ?? 0,
      recentCommunications: json["recent_communications"] ?? 0,
      communicationsByMode: modeMap,
      lastCommunicationDate: json["last_communication_date"] != null
          ? DateTime.parse(json["last_communication_date"])
          : null,
    );
  }
  final int totalCommunications;
  final int unansweredCommunications;
  final int recentCommunications; // Last 7 days
  final Map<CommunicationMode, int> communicationsByMode;
  final DateTime? lastCommunicationDate;

  Map<String, dynamic> toJson() {
    final Map<String, int> modeMap = {};
    communicationsByMode.forEach((mode, count) {
      modeMap[mode.value] = count;
    });

    return {
      "total_communications": totalCommunications,
      "unanswered_communications": unansweredCommunications,
      "recent_communications": recentCommunications,
      "communications_by_mode": modeMap,
      "last_communication_date": lastCommunicationDate?.toIso8601String(),
    };
  }

  // Get percentage of unanswered communications
  double get unansweredPercentage {
    if (totalCommunications == 0) return 0;
    return (unansweredCommunications / totalCommunications) * 100;
  }
}
